girder/api/v1/resource.py

```diff
@@ -139,6 +140,96 @@ def _getResourceModel(self, kind, funcName=None):
             raise RestException('Invalid resources format.')
         return model
 
+    def _lookupToken(self, token, parentType, parent):
+        """
+        Find a particular child resource by name or throw an exception.
+        :param token: the name of the child resource to find
+        :param parentType: the type of the parent to search
+        :param parent: the parent resource
+        :returns: the child resource
+        """
+
+        filterObject = {'name': token, 'parentId': parent['_id']}
+
+        # look for a folder
+        candidateChild = self.model('folder').findOne(filterObject)
+        if candidateChild is not None:
+            return candidateChild, 'folder'
+
+        # if folder not found, look for an item
+        filterObject['folderId'] = filterObject.pop('parentId')
+        candidateChild = self.model('item').findOne(filterObject)
+        if candidateChild is not None:
+            return candidateChild, 'item'
+
+        # if item not found, look for a file
+        filterObject['itemId'] = filterObject.pop('folderId')
+        candidateChild = self.model('file').findOne(filterObject)
+        if candidateChild is not None:
+            return candidateChild, 'file'
+
```

Comment:
We should add some conditions to the above code so that we don't do unnecessary lookups. Namely,

1. If parentType is "user" or "collection", we only need to lookup folders.
2. If parentType is "folder", we only need to lookup folders and items.
3. If parentType is "item", we only need to lookup files.

Additionally, when looking up folders, we should include `'parentCollection': parentType` in the filter dict.